code-monkey/src/agents/agent.py

```diff
@@ -1,78 +1,148 @@
 import os
+from typing import Set
+from langgraph.prebuilt import create_react_agent
+from langchain_core.messages import HumanMessage, SystemMessage
+from langchain_core.language_models.chat_models import BaseChatModel
+
 from tools.utils import (
     ask_user,
     show_diff,
 )
-from constants import get_src_dir, get_artifacts_dir
-from models import Model, get_model_service
+from constants import get_root_dir, get_artifacts_dir
+from models import parse_msn
 from .base_agent import BaseAgent
 from instrumentation import current_span, instrument
+from langgraph.checkpoint.aiosqlite import AsyncSqliteSaver
 
 class Agent(BaseAgent):
-    model: Model
+    model: BaseChatModel
+
+    config = {
+        "configurable": {"thread_id": "abc123"},
+    }
 
+    @instrument("Agent.__init__", ["msn"])
     def __init__(self, msn: str | None):
-        # TODO(toshok) these two steps should be collapsed and hidden back in the
-        # models/ directory (and replaced with something like this here:
-        #   self.model = construct_model(msn, self)
-        Service = get_model_service(msn)
-        self.model = Service(self, msn)
+        [ ChatModel, model, extra_flags ] = parse_msn(msn)
+
+        # a checkpointer + the thread_id below gives the model a way to save its
+        # state so we don't have to accumulate the messages ourselves.
+        memory = AsyncSqliteSaver.from_conn_string(":memory:")
```

Comment:
Oh man... These system dependencies are harsh. Is the sqlite installation part of the python package, or is it in setup.sh?

Reply:
it's part of the python package on macos:

```
% brew info python3
==> python@3.12: stable 3.12.4 (bottled)
Interpreted, interactive, object-oriented programming language
https://www.python.org/
Installed
/opt/homebrew/Cellar/python@3.12/3.12.4 (3,312 files, 66.3MB)
  Poured from bottle using the formulae.brew.sh API on 2024-06-30 at 16:09:02
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/p/python@3.12.rb
License: Python-2.0
==> Dependencies
Build: pkg-config ✔
Required: mpdecimal ✔, openssl@3 ✔, sqlite ✔, xz ✘
```


I presume it is on linux as well, but depends on the package manager's configuration I guess.